File: Week_09/java/src/_214_ShortestPalindrome.java
/**
 * 第一遍：2020/08/24周二 ✅
 * 第二遍：2020/08/25周三 ✅
 * 第三遍：2020/08/25周五 ✅
 * 第四遍：2020/06/13周六
 * todo: 使用二进制位的解法
 */
class _214_ShortestPalindrome {
    private static final int D = 256;
    private static final int Q = 1000000007;

    public String shortestPalindrome(String s) {
        if (s == null || s.isEmpty()) return s;
        long u = 0, v = 0, m = 1;
        int n = s.length(), r = 0;
        for (int i = 0; i < n; i++) {
            u = (u * D + s.charAt(i)) % Q;
            v = (v + m * s.charAt(i)) % Q;
            m = m * D % Q;
            if (u == v) {
                r = i;
            }
        }
        StringBuilder sb = new StringBuilder();
        for (int i = n - 1; i > r; i--) {
            sb.append(s.charAt(i));
        }
        sb.append(s);
        return sb.toString();
    }

    class SolutionKMP {
        public String shortestPalindrome(String s) {
            if (s == null || s.isEmpty()) return s;
            int n = s.length();
            int[] lps = new int[n + 1];
            lps[0] = -1;
            int j = 0;
            int i = 1;
            while (i < n) {
                if (s.charAt(j) == s.charAt(i)) {
                    lps[++i] = ++j;
                } else if (j != 0) {
                    j = lps[j];
                } else {
                    i++;
                }
            }
            j = 0;
            for (i = n - 1; i >= 0; i--) {
                while (j > 0 && s.charAt(j) != s.charAt(i)) {
                    j = lps[j];
                }
                if (s.charAt(j) == s.charAt(i)) {
                    j++;
                }
            }
            StringBuilder sb = new StringBuilder();
            for (i = n - 1; i >= j; i--) {
                sb.append(s.charAt(i));
            }
            sb.append(s);
            return sb.toString();
        }
    }
}

File: Week_09/kotlin/src/_214_ShortestPalindrome.kt
fun shortestPalindrome(s: String): String {
    val r = StringBuilder(s).reverse()
    for (i in s.indices) {
        if (s.startsWith(r.substring(i))) {
            return r.substring(0, i) + s
        }
    }
    return ""
}

private const val D = 256
private const val Q = 1000000007

class _214_RabinKarp {
    fun shortestPalindrome(s: String): String {
        var l = 0L
        var r = 0L
        var m = 1L
        var rMax = 0
        for (i in s.indices) {
            l = (l * D + s[i].toLong()) % Q
            r = (s[i].toLong() * m + r) % Q
            m = m * D % Q
            if (l == r) rMax = i
        }
        val sb = StringBuilder()
        for (i in s.lastIndex downTo rMax + 1) {
            sb.append(s[i])
        }
        return sb.append(s).toString()
    }
}

class _214_KMP {
    fun shortestPalindrome(s: String): String {
        val n = s.length
        val lps = IntArray(n + 1).apply { this[0] = -1 }
        var j = 0
        var i = 1
        while (i < n) {
            when {
                s[i] == s[j] -> lps[++i] = ++j
                j > 0 -> j = lps[j]
                else -> i++
            }
        }

        i = n - 1
        j = 0
        while (i >= 0) {
            when {
                s[i] == s[j] -> {
                    i--;j++
                }
                j > 0 -> j = lps[j]
                else -> i--
            }
        }

        val sb = StringBuilder()
        i = n - 1
        while (i >= j) {
            sb.append(s[i--])
        }
        sb.append(s)
        return sb.toString()
    }
}